parsers/CNDC.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+
+import json
+import re
+from datetime import datetime
+from logging import Logger, getLogger
+from zoneinfo import ZoneInfo
+
+from requests import Session
+
+from electricitymap.contrib.lib.models.event_lists import (
+    EventSourceType,
+    ProductionBreakdownList,
+    ProductionMix,
+    TotalProductionList,
+)
+from electricitymap.contrib.lib.types import ZoneKey
+
+tz_bo = ZoneInfo("America/La_Paz")
+
+INDEX_URL = "https://www.cndc.bo/gene/index.php"
+DATA_URL = "https://www.cndc.bo/gene/dat/gene.php?fechag={0}"
+SOURCE = "cndc.bo"
+
+
+def extract_xsrf_token(html):
+    """Extracts XSRF token from the source code of the generation graph page."""
+    return re.search(r'var ttoken = "([a-f0-9]+)";', html).group(1)
+
+
+def get_timestamp(query_date: datetime, hour: int) -> datetime:
+    return datetime(
+        year=query_date.year,
+        month=query_date.month,
+        day=query_date.day,
+        hour=hour,
+        tzinfo=tz_bo,
+    )
+
+
+def fetch_data(
+    session: Session | None = None, target_datetime: datetime | None = None
+) -> tuple[list[dict], datetime]:
+    if session is None:
+        session = Session()
+
+    if target_datetime is None:
+        target_datetime = datetime.now()
+    target_datetime = target_datetime.astimezone(tz_bo)
+    # Define actual and previous day (for midnight data).
+    formatted_dt = target_datetime.strftime("%Y-%m-%d")
+
+    # XSRF token for the initial request
+    xsrf_token = extract_xsrf_token(session.get(INDEX_URL).text)
+
+    resp = session.get(
+        DATA_URL.format(formatted_dt), headers={"x-csrf-token": xsrf_token}
+    )
+
+    hour_rows = json.loads(resp.text.replace("ï»¿", ""))["data"]
+    return hour_rows, target_datetime
+
+
+def parse_generation_forecast(
+    zone_key: ZoneKey, date: datetime, raw_data: list[dict], logger: Logger
+) -> TotalProductionList:
+    result = TotalProductionList(logger)
+    assert date.tzinfo == tz_bo
+    for hour_row in raw_data:
+        [hour, forecast, total, thermo, hydro, wind, solar, bagasse] = hour_row
+
+        # "hour" is one-indexed
+        timestamp = get_timestamp(query_date=date, hour=hour - 1)
```

Comment:
Not sure if you changed this part but I would prefer if it was called something like get_datetime instead as it actually returns a datetime.